volumes.c

```diff
@@ -92,19 +125,100 @@ void pv_volumes_empty(struct pv_state *s)
 	pv_log(INFO, "removed %d volumes", num_vol);
 }
 
-struct pv_volume* pv_volume_add(struct pv_state *s, char *name)
+struct pv_volume* pv_volume_add_with_disk(struct pv_state *s, char *name, char *disk)
 {
 	struct pv_volume *v = calloc(1, sizeof(struct pv_volume));
+	struct pv_disk *d, *tmp;
+	struct dl_list *disks = NULL;
 
 	if (v) {
 		v->name = strdup(name);
 		dl_list_init(&v->list);
 		dl_list_add_tail(&s->volumes, &v->list);
+		disks = &s->disks;
+
+		if (disk) {
+			dl_list_for_each_safe(d, tmp, disks, struct pv_disk, list) {
+				if (!strcmp(d->name, disk)) {
+					v->disk = d;
+					break;
+				}
+			}
+		}
+		else {
+			dl_list_for_each_safe(d, tmp, disks, struct pv_disk, list) {
+				if (d->def) {
+					v->disk = d;
+					break;
+				}
+			}
+		}
 	}
 
 	return v;
 }
 
+struct pv_volume* pv_volume_add(struct pv_state *s, char *name)
+{
+	return pv_volume_add_with_disk(s, name, NULL);
+}
+
+struct pv_disk* pv_disk_add(struct pv_state *s)
+{
+	struct pv_disk *d = calloc(1, sizeof(struct pv_disk));
+
+	if (d) {
+		dl_list_init(&d->list);
+		dl_list_add_tail(&s->disks, &d->list);
+	}
+
+	return d;
+}
+
+static int pv_volumes_mount_handler(struct pantavisor *pv, struct pv_volume *v, char *action)
+{
+	struct pv_disk *d = v->disk;
+	char *command = NULL;
+	int wstatus, ret = -1;
+
+	switch (d->type) {
+	case DISK_DM_CRYPT_CAAM:
+	command = malloc(sizeof(char) *
+		(strlen("/lib/pv/volmount/crypt caam %s %s /volumes/%s/%s %s") +
+		strlen(action) +
+		strlen(d->path) +
+		strlen(v->plat->name) +
+		strlen(v->name) +
+		strlen("&>> /storage/disks/log")) + 1);
+
+		sprintf(command, "/lib/pv/volmount/crypt caam %s %s /volumes/%s/%s &>> /storage/disks/log",
+				  action, d->path, v->plat->name, v->name);
+		pv_log(INFO, "command: %s", command);
+		break;
+	case DISK_DM_CRYPT_DCP:
+	case DISK_DIR:
+	case DISK_DM_CRYPT_VERSATILE:
+	case DISK_EXT4_UUID:
+	case DISK_EXT4:
+	case DISK_UNKNOWN:
+	default:
+		return -ENOTSUP;
+	}
+
+	tsh_run(command, 1, &wstatus);
```

Comment:
see the above comment about tsh_run_output

Reply:
Done, thanks